xptifw/src/xpti_trace_framework.cpp

```diff
@@ -0,0 +1,1204 @@
+    bool Success = StreamCBs.find(Acc, TraceType);
+#else
+    auto Acc = StreamCBs.find(TraceType);
+    bool Success = (Acc != StreamCBs.end());
+#endif
+    if (Success) {
+      for (auto &Ele : Acc->second) {
+        if (Ele.second == cbFunc) {
+          if (Ele.first) { // Already here and active
+                           // unregister, since delete and simultaneous
+                           // iterations by other threads are unsafe
+            Ele.first = false;
+            // releases the accessor
+            return xpti::result_t::XPTI_RESULT_SUCCESS;
+          } else {
+            // releases the accessor
+            return xpti::result_t::XPTI_RESULT_DUPLICATE;
+          }
+        }
+      }
+    }
+    //  Not here, so nothing to unregister
+    return xpti::result_t::XPTI_RESULT_NOTFOUND;
+  }
+
+  xpti::result_t unregisterStream(uint8_t StreamID) {
+    // If there are no callbacks registered for the requested stream ID, we
+    // return not found
+#ifndef XPTI_USE_TBB
+    std::lock_guard<std::mutex> Lock(MCBsLock);
+#endif
+    if (MCallbacksByStream.count(StreamID) == 0)
+      return xpti::result_t::XPTI_RESULT_NOTFOUND;
+
+    auto &StreamCBs = MCallbacksByStream[StreamID]; // thread-safe
+    // Disable all callbacks registered for the stream represented by StreamID
+    for (auto &Item : StreamCBs) {
+      for (auto &Ele : Item.second) {
+        Ele.first = false;
+      }
+    }
+    //  Return success
+    return xpti::result_t::XPTI_RESULT_SUCCESS;
+  }
+
+  xpti::result_t notifySubscribers(uint16_t StreamID, uint16_t TraceType,
+                                   xpti::trace_event_data_t *Parent,
+                                   xpti::trace_event_data_t *Object,
+                                   uint64_t InstanceNo, const void *UserData) {
+    {
+#ifndef XPTI_USE_TBB
+      std::lock_guard<std::mutex> Lock(MCBsLock);
+#endif
+      cb_t &Stream = MCallbacksByStream[StreamID]; // Thread-safe
+#ifdef XPTI_USE_TBB
+      cb_t::const_accessor Acc; // read-only accessor
+      bool Success = Stream.find(Acc, TraceType);
+#else
+      auto Acc = Stream.find(TraceType);
+      bool Success = (Acc != Stream.end());
+#endif
+
+      if (Success) {
+        // Go through all registered callbacks and invoke them
+        for (auto &Ele : Acc->second) {
+          if (Ele.first)
+            (Ele.second)(TraceType, Parent, Object, InstanceNo, UserData);
+        }
+      }
+    }
+#ifdef XPTI_STATISTICS
+    auto &Counter = MStats[TraceType];
+    {
+#ifdef XPTI_USE_TBB
+      tbb::spin_mutex::scoped_lock Lock(MStatsLock);
+#else
+      std::lock_guard<std::mutex> Lock(MStatsLock);
+#endif
+      Counter++;
+    }
+#endif
+    return xpti::result_t::XPTI_RESULT_SUCCESS;
+  }
+
+  void printStatistics() {
+#ifdef XPTI_STATISTICS
+    printf("Notification statistics:\n");
+    for (auto &s : MStats) {
+      printf("%19s: [%llu] \n",
+             stringify_trace_type((xpti_trace_point_type_t)s.first).c_str(),
```

Comment:
At least with `std::cout` there is no need to go back to a C string :-)

Reply:
:-) Agreed, but formatting string in <iostream> is too verbose - and printfs are used only when someone enables usage statistics